Add a test for decompositions that decomposes all the operations as much as possible

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92243
* #92004
* __->__ #87182
* #87183
* #92003

This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc @ezyang @mruberry @ngimel @Lezcano @fdrocha @peterbell10

lezcano committed:
Add a test for decompositions that decomposes all the operations as much as possible

This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano commented:
Because it takes quite long. Now, it found SO MANY bugs (all this stack is bug fixes for these bugs), that one could make the case that this is actually the way we want to test things. I'll leave it up to you.

ezyang commented on test/test_decomp.py:
```diff
@@ -412,6 +427,8 @@ def _torch_dispatch(self, func, types, args=(), kwargs=None):
                 # Stuff we shouldn't bother testing
                 # (TODO: remove detach from the decomp table?)
+                # N.b. Testing in-place ops would need dedicated logic
+                in_place = func.name()[-1] == '_'
```
@anjali411 don't we have a more direct way of querying if an OpOverload is inplace now?

ezyang commented on test/test_decomp.py:
```diff
@@ -422,7 +439,7 @@ def _torch_dispatch(self, func, types, args=(), kwargs=None):
                     torch.ops.aten.randn.default,
                     torch.ops.aten.native_dropout.default,
-                ] or any_unsupported(args, kwargs):
+                ] or any_unsupported(args, kwargs) or in_place:
```
doesn't this change the behavior of non-recursive decomps too?

lezcano replied:
Niet. in-place ops were not being tested, before, and those that were, happened to pass this by chance. There is no logic in this function to handle in-place functions.

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano commented:
Actually, I'm tempted to make this the default behaviour. It's slow, but it's not slower than most tests in opinfos really (think gradtest). Also, the wins from finding these issues early are massive. What do you think @ezyang?

ezyang commented:
I am for it

ezyang commented:
Get rid of the old test whole youre at it

Chillee commented:
I'm also for it. I actually thought this was how we usually tested :P

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

This test is slow, so we mark it with `slowTest`

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc ezyang mruberry ngimel Lezcano fdrocha peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc ezyang mruberry ngimel Lezcano fdrocha peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc ezyang mruberry ngimel Lezcano fdrocha peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc ezyang mruberry ngimel Lezcano fdrocha peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc ezyang mruberry ngimel Lezcano fdrocha peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc ezyang mruberry ngimel Lezcano fdrocha peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc ezyang mruberry ngimel Lezcano fdrocha peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc ezyang mruberry ngimel Lezcano fdrocha peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc ezyang mruberry ngimel Lezcano fdrocha peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Add a test for decompositions that decomposes all the operations as much as possible"


This will enable a more thorough testing of the decompositions than the
one just provided by OpInfos.

cc ezyang mruberry ngimel Lezcano fdrocha peterbell10

[ghstack-poisoned]
